core/domain/exp_jobs_one_off.py

```diff
@@ -216,68 +256,133 @@ def map(item):
+                            output_values = '%s %s: %s' % (
+                                item.id, state_name, rule_input)
+                            yield (validity, output_values.encode('utf-8'))
+
+    @staticmethod
+    def reduce(key, values):
+        if key.startswith('Valid'):
+            yield (key, values[:VALID_MATH_INPUTS_YIELD_LIMIT])
+        else:
+            if key == 'Invalid' and len(values) != 0:
+                yield ('ERROR', (
+                    'There are some invalid inputs that need to be resolved '
+                    'before running the upgradation job. The invalid inputs '
+                    'are as follows:'))
+            yield (key, values)
+
+
+class MathExpressionUpgradationOneOffJob(jobs.BaseMapReduceOneOffJobManager):
+    """Job that produces a list of explorations that use the MathExpressionInput
+    along with the validity and type (expression/equation) of the inputs present
+    in the exploration.
+
+    This validation is done by the 'is_valid_math_expression' or the
+    'is_valid_math_equation' function present in schema_utils.py.
+    """
+
+    @classmethod
+    def entity_classes_to_map_over(cls):
+        return [exp_models.ExplorationModel]
+
+    @staticmethod
+    def map(item):
+        is_valid_math_expression = schema_utils.get_validator(
+            'is_valid_math_expression')
+        is_valid_math_equation = schema_utils.get_validator(
+            'is_valid_math_equation')
+        ltt = latex2text.LatexNodes2Text()
+
+        if not item.deleted:
+            exploration = exp_fetchers.get_exploration_from_model(item)
+            for state_name, state in exploration.states.items():
+                if state.interaction.id == 'MathExpressionInput':
+                    new_answer_groups = []
+                    new_interaction_id = ''
+                    for group in state.interaction.answer_groups:
+                        new_answer_group = copy.deepcopy(group)
+                        for rule_spec in new_answer_group.rule_specs:
+                            rule_input = ltt.latex_to_text(
+                                rule_spec.inputs['x'])
+
+                            rule_input = clean_math_expression(rule_input)
+
+                            validity = 'Invalid'
+                            if is_valid_math_expression(rule_input):
+                                validity = 'Valid Algebraic Expression'
+                                new_interaction_id = 'AlgebraicExpressionInput'
+                            elif is_valid_math_expression(
+                                    rule_input, algebraic=False):
+                                validity = 'Valid Numeric Expression'
+                                new_interaction_id = 'NumericExpressionInput'
+                            elif is_valid_math_equation(rule_input):
+                                validity = 'Valid Math Equation'
+                                new_interaction_id = 'MathEquationInput'
```

Comment:
You still need to raise/yield/etc. if an error is discovered, and handle that case safely, providing enough information to the admin such that they can take manual measures if needed or re-run the job again.

Do not leave anything to chance in a migration. Mistakes happen, so program defensively and do not assume anything.

Reply:
Done.

Reply:
I've handled this now. If any of the inputs are categorized as Invalid, an error will be generated and that state won't be upgraded.